toolchain/check/member_access.cpp

```diff
@@ -502,70 +527,79 @@ static auto PerformActionHelper(Context& context, SemIR::LocId loc_id,
-  llvm::SmallVector<LookupScope> lookup_scopes;
-  if (!AppendLookupScopesForConstant(context, loc_id, base_type_const_id,
-                                     &lookup_scopes)) {
-    // The base type is not a name scope. Try some fallback options.
-    if (auto struct_type = context.insts().TryGetAs<SemIR::StructType>(
-            context.constant_values().GetInstId(base_type_const_id))) {
-      // TODO: Do we need to optimize this with a lookup table for O(1)?
-      for (auto [i, field] : llvm::enumerate(
-               context.struct_type_fields().Get(struct_type->fields_id))) {
-        if (name_id == field.name_id) {
-          // TODO: Model this as producing a lookup result, and do instance
-          // binding separately. Perhaps a struct type should be a name scope.
-          return GetOrAddInst<SemIR::StructAccess>(
-              context, loc_id,
-              {.type_id =
-                   context.types().GetTypeIdForTypeInstId(field.type_inst_id),
-               .struct_id = base_id,
-               .index = SemIR::ElementIndex(i)});
-        }
-      }
-      if (required) {
-        CARBON_DIAGNOSTIC(QualifiedExprNameNotFound, Error,
-                          "type {0} does not have a member `{1}`", TypeOfInstId,
-                          SemIR::NameId);
-        context.emitter().Emit(loc_id, QualifiedExprNameNotFound, base_id,
-                               name_id);
-        return SemIR::ErrorInst::InstId;
-      } else {
-        return SemIR::InstId::None;
+  {
+    // If `base_type_id` is a facet, we don't know its eventual type yet, but we
+    // don't produce a symbolic instruction to do the name lookup later. We want
+    // to do that lookup into the scope of the facet's FacetType, so we extract
+    // that here.
+    auto lookup_type_id = ExtractFacetTypeForFacet(context, base_type_id);
+    auto lookup_type_const_id = context.types().GetConstantId(lookup_type_id);
+
+    llvm::SmallVector<LookupScope> lookup_scopes;
+    if (AppendLookupScopesForConstant(context, loc_id, lookup_type_const_id,
+                                      &lookup_scopes)) {
+      // Perform lookup into the base type.
+      auto member_id = LookupMemberNameInScope(
+          context, loc_id, base_id, name_id, lookup_type_const_id,
+          lookup_scopes,
+          /*lookup_in_type_of_base=*/true, /*required=*/required);
+
+      if (perform_instance_binding) {
+        // Perform instance binding if we found an instance member.
+        member_id = PerformInstanceBinding(context, loc_id, base_id, member_id);
       }
-    }
 
-    if (base_type_id != SemIR::ErrorInst::TypeId) {
-      CARBON_DIAGNOSTIC(QualifiedExprUnsupported, Error,
-                        "type {0} does not support qualified expressions",
-                        TypeOfInstId);
-      context.emitter().Emit(loc_id, QualifiedExprUnsupported, base_id);
+      return member_id;
     }
-    return SemIR::ErrorInst::InstId;
```

Comment:
I inverted the `if` above, and moved this whole block out of the if, but it's unchanged otherwise.

Then the code below moved into the `if` with some slight changes.

In this case context helps make the diff make more sense.